autox/environment/nncf/quantizer_tracing.py

```diff
@@ -0,0 +1,511 @@
+                            #  invalidated if input to multiple gemm
+                            if len(list(g.get_successors(innodekey))) != 1:
+                                raise ValueError("Model Input fans in to multiple nodes, assumption breaks, pls debug")
+                            
+                            input_module_nodekey = list(g.get_successors(innodekey))[0]
+                            IQId = InputQuantizerId(g.get_nx_node_by_key(input_module_nodekey)['op_exec_context'].input_agnostic)
+                            
+                            if NWQId_exist_Qctrl(IQId, quantization_controller):
+                                gemm_quantizer_dict[gemm]['NWQ'].append((innodekey, str(IQId)))
+                                print("==> Found NW Quantizer | {}".format(str(IQId)))
+                            else:
+                                raise ValueError("IQId not found, Pls Debug")
+                        else:
+                            NWQId=get_non_weight_quantizer_id(g, innode)
+                            
+                            if NWQId_exist_Qctrl(NWQId, quantization_controller):
+                                gemm_quantizer_dict[gemm]['NWQ'].append((innodekey, str(NWQId)))
+                                print("==> Found NW Quantizer | {}".format(str(NWQId)))
+                            else:
+                                raise ValueError("NWQId not found, Pls Debug")
+                # [End OF] Recursively trace until a non-weight quantizer ====================
+    
+    # Bookkeep GEMM Quantize Nodekey
+    for gemm_scope, gemm_dict in gemm_quantizer_dict.items():
+        wt_quantize_nodekey, wqid = gemm_dict['WQ']
+        
+        child_nx_nodes = list(g.get_successors(wt_quantize_nodekey))
+        if len(child_nx_nodes) != 1:
+            raise ValueError("Child of weight quantize node should be GEMM node, assumption Breaks, pls debug")
+        
+        gemm_nx_nodekey = child_nx_nodes[0]        
+        gemm_quantizer_dict[gemm_scope]['nx_nodekey']=gemm_nx_nodekey
+
+    print("\n")
+    return gemm_quantizer_dict
+
+def get_untagged_quantizer(quantization_controller, gemm_quantizer_dict):
+    # Find untagged quantizer
+    QId_list=list(map(str,quantization_controller.all_quantizations.keys()))
+
+    gemm_tagged_quantizer_tuple_list = [] 
+    for gemm_scope, gemm_dict in gemm_quantizer_dict.items():
+        gemm_tagged_quantizer_tuple_list.append(gemm_dict["WQ"])
+        gemm_tagged_quantizer_tuple_list.extend(gemm_dict["NWQ"])
+
+    gemm_tagged_quantizer_nx_nodekey_list=list(map(lambda x:x[0], gemm_tagged_quantizer_tuple_list))
+    gemm_tagged_QId_list=list(map(lambda x:x[1], gemm_tagged_quantizer_tuple_list))
+
+    untagged_QId_list = list(set(QId_list)-set(gemm_tagged_QId_list))
+    return untagged_QId_list
+
+# MAIN ====================================================================================
+if __name__ == '__main__':
```

Comment:
I believe the library code shouldn't be called. It should be moved to function or class

Reply:
Low priority. Please consider removing this code.

Reply:
not applicable anymore. quantizer_tracing.py has been removed and functionality has been replaced.